### PR TITLE
Vectorize copysign on CPU

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50438 Support JIT shape analysis for div with rounding_mode argument
* #55225 grad_fn getter for optional strings
* #51989 Sparse support for division rounding_mode argument
* #55380 Vectorize floating point floor_divide
* **#51792 Vectorize copysign on CPU**

Differential Revision: [D27769007](https://our.internmc.facebook.com/intern/diff/D27769007)

### CODE REVIEW

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 1]
```diff
@@ -43,12 +43,13 @@ namespace at {
...
 namespace {
-// at::Half should be treated as floating point
+// at::Half and at::BFloat16 should be treated as floating point
 template <typename T>
 struct is_floating_point:
     std::integral_constant<bool,
       std::is_floating_point<T>::value ||
-      std::is_same<T, at::Half>::value> {
+      std::is_same<T, at::Half>::value ||
+      std::is_same<T, at::BFloat16>::value> {
```
Comment (mruberry) [inline aten/src/ATen/cpu/vec256/vec256_base.h:52]: Thanks for adding this.

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 2]
```diff
@@ -316,6 +317,13 @@ struct Vec256 {
     }
     return ret;
   }
+  Vec256<T> copysign(const Vec256<T> &sign) const {
+    Vec256<T> ret;
+    for (int64_t i = 0; i < size(); i++) {
```
Comment (mruberry) [inline aten/src/ATen/cpu/vec256/vec256_base.h:322]: type `i` like `size()`

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h [hunk 3]
```diff
@@ -324,6 +324,16 @@ template <> class Vec256<float> {
     }
     return loadu(tmp);
   }
+  Vec256<float> copysign(const Vec256<float> &sign) const {
+    __at_align32__ float tmp[size()];
+    __at_align32__ float tmp_sign[size()];
+    store(tmp);
+    sign.store(tmp_sign);
+    for (int64_t i = 0; i < size(); i++) {
```
Comment (mruberry) [inline aten/src/ATen/cpu/vec256/vec256_float_neon.h:332]: type `i` like `size`

**Comment by mruberry** [comment]
Overall this is great and it'll be awesome to have vectorized floor_divide.

So this is also adding CPU vectorized copysign? cc @malfet who recently looked at CPU copysign issues.

How does this relate to the existing CPU copysign code here: 

https://github.com/pytorch/pytorch/blob/b880854f31e9ef48bda2b48bc9e5b6961aa8e8bb/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L22

and here:

https://github.com/pytorch/pytorch/blob/b880854f31e9ef48bda2b48bc9e5b6961aa8e8bb/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L935

Maybe we should separate out the copysign changes into a PR before updating floor_divide?

**Comment by peterbell10** [comment]
> How does this relate to the existing CPU `copysign` code here

Now you point it out, I've moved the `copysign` implementations into `c10/util` so it can be used by the vectorized code as well. I noticed there was a `BFloat-math.h` but not an equivalent for `Half`, so created that.

>Maybe we should separate out the copysign changes into a PR before updating floor_divide?

This PR is now for `copysign`-only and `floor_divide` code was moved to #55380.

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 4]
```diff
@@ -316,6 +318,13 @@ struct Vec256 {
     }
     return ret;
   }
+  Vec256<T> copysign(const Vec256<T> &sign) const {
+    Vec256<T> ret;
+    for (int i = 0; i < size(); i++) {
```
Comment (mruberry) [inline aten/src/ATen/cpu/vec256/vec256_base.h:323]: type `i` like `size()`

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h [hunk 5]
```diff
@@ -319,11 +319,21 @@ template <> class Vec256<float> {
     __at_align32__ float tmp_exp[size()];
     store(tmp);
     exp.store(tmp_exp);
-    for (int64_t i = 0; i < size(); i++) {
+    for (int i = 0; i < size(); i++) {
```
Comment (mruberry) [inline aten/src/ATen/cpu/vec256/vec256_float_neon.h:322]: type `i` like `size()`:

```
auto i = decltype(size()){0}
```

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 6]
```diff
@@ -2995,7 +2995,7 @@ void quantize_tensor_per_tensor_affine_sub_byte_cpu(
       const auto elem_per_byte = CHAR_BIT / bit_width;
       for (int i = 0; i < numel; ++i) {
         float inv_scale = scale == 0 ? 1.0f : 1.0f / scale;
-        int qvalue = lrintf(std::nearbyint(rdata[i] * inv_scale) + zero_point);
+        int64_t qvalue = lrintf(std::nearbyint(rdata[i] * inv_scale) + zero_point);
```
Comment (mruberry) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:2998]: Why this change?

From a style perspective this is why I prefer auto.

File: aten/src/ATen/native/quantized/fake_quant_per_tensor_affine.cpp [hunk 7]
```diff
@@ -108,7 +108,7 @@ int64_t _get_zero_point_from_tensor(
     bool is_forward) {
   float zero_point_fp = zero_point[0].item<float>();
   zero_point_fp = is_forward ? std::nearbyint(zero_point_fp) : zero_point_fp + 0.5f;
-  float zero_point_clamped = std::min(std::max(zero_point_fp, quant_min), quant_max);
+  float zero_point_clamped = std::min(std::max(zero_point_fp, float(quant_min)), float(quant_max));
```
Comment (mruberry) [inline aten/src/ATen/native/quantized/fake_quant_per_tensor_affine.cpp:111]: Why this change?

File: c10/util/Half-math.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <c10/util/Half.h>
+#include <c10/util/math_compat.h>
+
+namespace std {
+
+/// Used by vec256<c10::Half>::map
+inline c10::Half acos(c10::Half a) { return std::acos(float(a));}
```
Comment (mruberry) [inline c10/util/Half-math.h:9]: I'm not sure what the history of half math is. @ngimel, do you know?

File: aten/src/ATen/cpu/vec256/vec256_base.h (hunk 4 above)
Comment (peterbell10) [reply aten/src/ATen/cpu/vec256/vec256_base.h:323]: I've already changed it. `size()` is `int`.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 6 above)
Comment (peterbell10) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:2998]: Same as https://github.com/pytorch/pytorch/pull/51792#discussion_r607901624

File: aten/src/ATen/native/quantized/fake_quant_per_tensor_affine.cpp (hunk 7 above)
Comment (peterbell10) [reply aten/src/ATen/native/quantized/fake_quant_per_tensor_affine.cpp:111]: Same as https://github.com/pytorch/pytorch/pull/51792#discussion_r607901624

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 9]
```diff
@@ -71,7 +52,8 @@ void add_clamp_kernel(TensorIterator& iter, const Scalar& alpha_scalar, const Sc
     auto max_vec = Vec256<scalar_t>(max_scalar);
     cpu_kernel_vec(iter,
       [=](scalar_t a, scalar_t b) __ubsan_ignore_undefined__ -> scalar_t {
-        return std::min(max_scalar, std::max(min_scalar, a + alpha * b));
+        scalar_t x = a + alpha * b;
+        return std::min(max_scalar, std::max(min_scalar, x));
```
[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:56]: There's been a recently fixed issue for this https://github.com/pytorch/pytorch/issues/55613

File: c10/util/Half-math.h (hunk 8 above)
Comment (ngimel) [reply c10/util/Half-math.h:9]: No, I don't. I once wondered how math overloads work, but didn't get a full picture, I wonder why it's necessary, shouldn't there be an implicit Half->float conversion (that's defined)? I thought that's how std math functions work in cuda kernels.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 10]
```diff
@@ -201,7 +183,7 @@ void div_floor_kernel(TensorIteratorBase& iter) {
                 floordiv += scalar_t(1.0);
               }
             } else {
-              floordiv = copysign(scalar_t(0), a / b);
+              floordiv = std::copysign(scalar_t(0), a / b);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:186]: You can not make this change, it will cause compiler ICE on aarch64 Ubuntu-18.04

File: aten/src/ATen/cpu/vec256/vec256_base.h (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/cpu/vec256/vec256_base.h:53]: Can it be put into separate PR (as it have nothing to do with vectorization goal)

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 11]
```diff
@@ -934,9 +916,13 @@ void heaviside_kernel(TensorIterator& iter) {
 
 void copysign_kernel(TensorIterator& iter) {
   AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, iter.common_dtype(), "copysign_cpu", [&]() {
-    cpu_kernel(iter, [](scalar_t a, scalar_t b) -> scalar_t {
-        return copysign(a, b);
-    });
+    cpu_kernel_vec(iter,
+      [](scalar_t a, scalar_t b) -> scalar_t {
+        return std::copysign(a, b);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:921]: Again, we can not use std::copysign otherwise it will cause an ICE

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 6 above)
Comment (malfet) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:2998]: This is already done in https://github.com/pytorch/pytorch/pull/55638

File: aten/src/ATen/native/quantized/fake_quant_per_tensor_affine.cpp (hunk 7 above)
Comment (malfet) [reply aten/src/ATen/native/quantized/fake_quant_per_tensor_affine.cpp:111]: Please use `static_cast<float>(quant_min)` (and if I'm not mistaken, this change is already done in #https://github.com/pytorch/pytorch/pull/55638 )

File: c10/util/Half-math.h [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+inline c10::Half sin(c10::Half a) { return std::sin(float(a));}
+inline c10::Half tan(c10::Half a) { return std::tan(float(a));}
+inline c10::Half tanh(c10::Half a) { return std::tanh(float(a));}
+inline c10::Half trunc(c10::Half a) { return std::trunc(float(a));}
+inline c10::Half lgamma(c10::Half a) { return std::lgamma(float(a));}
+inline c10::Half sqrt(c10::Half a) { return std::sqrt(float(a));}
+inline c10::Half rsqrt(c10::Half a) { return 1.0 / std::sqrt(float(a));}
+inline c10::Half abs(c10::Half a) { return std::abs(float(a));}
+inline c10::Half min(c10::Half a, c10::Half b) { return std::min(float(a), float(b));}
+inline c10::Half max(c10::Half a, c10::Half b) { return std::max(float(a), float(b));}
```
Comment (malfet) [inline c10/util/Half-math.h:33]: Please don't do that, it will reintroduce #55613, but for half instead of bfloat.
Instead, re-implement `operator<` and `operator>` for those types

**Review by malfet — CHANGES_REQUESTED** [review]
Please rebase against ToT to pick up fix for #55613
Also, do not introduce any new `std::min`,`std::max`  overrides for types that have default constructor, otherwise it can lead to unexpected sideeffects.

File: c10/util/Half-math.h (hunk 12 above)
Comment (mruberry) [reply c10/util/Half-math.h:33]: I'd also prefer we limit changes to half support as much as possible in this PR.

I'm guess what needs to be done can be done here?

https://github.com/pytorch/pytorch/blob/master/c10/util/Half-inl.h

File: aten/src/ATen/cpu/vec256/vec256_float_neon.h (hunk 5 above)
Comment (peterbell10) [reply aten/src/ATen/cpu/vec256/vec256_float_neon.h]: I really don't like `decltype(size())` so added a `using size_type` to `Vec256` that I can use instead.

File: c10/util/Half-math.h (hunk 8 above)
Comment (peterbell10) [reply c10/util/Half-math.h]: I assumed `BFloat-math.h` was an accepted pattern, but if that's not true then I've just removed it and focused on `copysign`.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 10 above)
Comment (peterbell10) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Renamed to `c10::` for clarity.

File: c10/util/Half-math.h (hunk 12 above)
Comment (peterbell10) [reply c10/util/Half-math.h]: Limited scope to just `copysign`.

**Comment by peterbell10** [comment]
PTAL @malfet, @mruberry

**Review by mruberry — COMMENTED** [review]
This looks OK to me but not an expert on copysign or c10/util. @malfet and @ezyang, would you two like to take a look?

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Actually, going to wait on @malfet, as I don't quite understand the ICE concerns.

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
Looks good to me, thank you for the updates

**Comment by mruberry** [comment]
Thanks for taking a look, @ezyang and @malfet!

**Review by mruberry — APPROVED** [review]
